File: user-service/src/test/java/com/example/user/controller/UserControllerTest.java
package com.example.user.controller;

import com.example.user.model.User;
import com.example.user.service.UserService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

class UserControllerTest {
    
    @Mock
    private UserService userService;
    
    @InjectMocks
    private UserController userController;
    
    @BeforeEach
    void setUp() {
        MockitoAnnotations.openMocks(this);
    }
    
    @Test
    void shouldCreateUserSuccessfully() {
        // Given
        User user = new User("testuser", "test@example.com", "password");
        User savedUser = new User("testuser", "test@example.com", "password");
        savedUser.setId(1L);
        
        when(userService.createUser(any(User.class))).thenReturn(savedUser);
        
        // When
        ResponseEntity<User> response = userController.createUser(user);
        
        // Then
        assertEquals(HttpStatus.CREATED, response.getStatusCode());
        assertNotNull(response.getBody());
        assertEquals(1L, response.getBody().getId());
        assertEquals("testuser", response.getBody().getUsername());
        
        verify(userService).createUser(any(User.class));
    }
    
    @Test
    void shouldReturnConflictWhenUserCreationFails() {
        // Given
        User user = new User("testuser", "test@example.com", "password");
        
        when(userService.createUser(any(User.class))).thenThrow(new RuntimeException("User already exists"));
        
        // When
        ResponseEntity<User> response = userController.createUser(user);
        
        // Then
        assertEquals(HttpStatus.CONFLICT, response.getStatusCode());
        verify(userService).createUser(any(User.class));
    }
    
    @Test
    void shouldGetUserByIdSuccessfully() {
        // Given
        Long id = 1L;
        User user = new User("testuser", "test@example.com", "password");
        user.setId(id);
        
        when(userService.getUserById(id)).thenReturn(Optional.of(user));
        
        // When
        ResponseEntity<User> response = userController.getUserById(id);
        
        // Then
        assertEquals(HttpStatus.OK, response.getStatusCode());
        assertNotNull(response.getBody());
        assertEquals(id, response.getBody().getId());
        assertEquals("testuser", response.getBody().getUsername());
        
        verify(userService).getUserById(id);
    }
    
    @Test
    void shouldReturnNotFoundWhenUserDoesNotExist() {
        // Given
        Long id = 1L;
        
        when(userService.getUserById(id)).thenReturn(Optional.empty());
        
        // When
        ResponseEntity<User> response = userController.getUserById(id);
        
        // Then
        assertEquals(HttpStatus.NOT_FOUND, response.getStatusCode());
        verify(userService).getUserById(id);
    }
}

File: user-service/src/test/kotlin/com/example/user/controller/UserControllerTest.kt
package com.example.user.controller

import com.example.user.model.User
import com.example.user.service.UserService
import io.mockk.*
import io.mockk.impl.annotations.MockK
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import java.util.*

class UserControllerTest {
    
    @MockK
    private lateinit var userService: UserService
    
    private lateinit var userController: UserController
    
    @BeforeEach
    fun setUp() {
        MockKAnnotations.init(this)
        userController = UserController(userService)
    }
    
    @Test
    fun `should create user successfully`() {
        // Given
        val user = User(username = "testuser", email = "test@example.com", password = "password")
        val savedUser = User(id = 1, username = "testuser", email = "test@example.com", password = "password")
        every { userService.createUser(any()) } returns savedUser
        
        // When
        val response = userController.createUser(user)
        
        // Then
        assert(response.statusCode == HttpStatus.CREATED)
        assert(response.body == savedUser)
        verify { userService.createUser(match { it.username == "testuser" && it.email == "test@example.com" }) }
    }
    
    @Test
    fun `should return conflict when user creation fails due to duplicate`() {
        // Given
        val user = User(username = "testuser", email = "test@example.com", password = "password")
        every { userService.createUser(any()) } throws RuntimeException("User already exists")
        
        // When
        val response = userController.createUser(user)
        
        // Then
        assert(response.statusCode == HttpStatus.CONFLICT)
        verify { userService.createUser(any()) }
    }
    
    @Test
    fun `should get user by ID successfully`() {
        // Given
        val id = 1L
        val user = User(id = id, username = "testuser", email = "test@example.com", password = "password")
        every { userService.getUserById(id) } returns Optional.of(user)
        
        // When
        val response = userController.getUserById(id)
        
        // Then
        assert(response.statusCode == HttpStatus.OK)
        assert(response.body == user)
        verify { userService.getUserById(id) }
    }
    
    @Test
    fun `should return not found when user does not exist`() {
        // Given
        val id = 1L
        every { userService.getUserById(id) } returns Optional.empty()
        
        // When
        val response = userController.getUserById(id)
        
        // Then
        assert(response.statusCode == HttpStatus.NOT_FOUND)
        verify { userService.getUserById(id) }
    }
    
    @Test
    fun `should delete user successfully`() {
        // Given
        val id = 1L
        every { userService.deleteUser(id) } returns true
        
        // When
        val response = userController.deleteUser(id)
        
        // Then
        assert(response.statusCode == HttpStatus.NO_CONTENT)
        verify { userService.deleteUser(id) }
    }
    
    @Test
    fun `should return not found when deleting non-existent user`() {
        // Given
        val id = 1L
        every { userService.deleteUser(id) } returns false
        
        // When
        val response = userController.deleteUser(id)
        
        // Then
        assert(response.statusCode == HttpStatus.NOT_FOUND)
        verify { userService.deleteUser(id) }
    }
}